### PR TITLE
Fixes an issue with mixed types in a subsequence

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/2195

### CODE REVIEW

File: src/org/exist/xquery/PathExpr.java [hunk 1]
```diff
@@ -297,7 +297,7 @@ public Sequence eval(Sequence contextSequence, final Item contextItem) throws XP
 
             final boolean allowMixedNodesInReturn;
             if(prev != null) {
-                allowMixedNodesInReturn = prev.allowMixedNodesInReturn() | expr.allowMixedNodesInReturn();
+                allowMixedNodesInReturn = prev.allowMixedNodesInReturn() /*| expr.allowMixedNodesInReturn()*/;
```
Comment (line-o) [inline src/org/exist/xquery/PathExpr.java]: @wolfgangmm it is unclear why this line needed to be changed for the test to succeed.

Comment (wolfgangmm) [reply src/org/exist/xquery/PathExpr.java]: @adamretter running the test on my install, above comment was not needed. The test passed without it. 

I did not contribute to the code calling allowMixedNodesInReturn, but I can see some sense in it. It's not vital though. I would thus say, if your test passes without commenting out the test above, we should leave PathExpr untouched. If it gets in your way though, we need to reconsider.

**Comment by adamretter** [comment]
@juri not intentionally!

File: src/org/exist/xquery/PathExpr.java (hunk 1 above)
Comment (adamretter) [reply src/org/exist/xquery/PathExpr.java]: @wolfgangmm Sure the contributed test is fine. But one of the existing tests in the test suite fails without commenting that out (due to my override in `Function.java`).

**Comment by juri** [comment]
@adamretter wrong person in the mention, I think…

File: src/org/exist/xquery/PathExpr.java (hunk 1 above)
Comment (wolfgangmm) [reply src/org/exist/xquery/PathExpr.java]: Ok, so the error expected by `XQueryTest#pathOperatorContainingNodesAndNonNodes()` is in line with what saxon reports. I wonder why you had to overwrite `Function#allowMixedNodesInReturn()`? It does not seem related to the original issue.

Comment (wolfgangmm) [reply src/org/exist/xquery/PathExpr.java]: Discard the question, I see it. This is quite a puzzle ...

**Comment by wolfgangmm** [comment]
@adamretter ok, after some hours I think I found the correct fix. Please try to change `SubSequence.getItemType()` to return the correct type of the underlying sequence:

```java
@Override
    public int getItemType() {
        return sequence.getItemType();
    }
```

With this in place you do not need to overwrite `Function.allowMixedNodesInReturn()` or `PathExpr`. Tests should pass. If this fails lazy evaluation, we will need another way to determine the expected type.

**Comment by adamretter** [comment]
@wolfgangmm So I don't think that will work for the very common `ValueSequence` which will still return `Type.ITEM`.

Is there anything wrong with my proposed changes. No tests fail. Are there things that you know that this breaks? If so, we should start by adding tests for those things...

**Comment by wolfgangmm** [comment]
`ValueSequence` does compute its type, so should be fine.

You wanted me to look into this because you had to comment out the check in `PathExpr`. This was originally added by yourself, so I assume you had a reason to do so. I have thus tried to find a solution which does not require the changes to `Function` and `PathExpr`.

All in all returning the correct type seems a more correct approach.

**Comment by adamretter** [comment]
@wolfgangmm Okay, I had no idea I added that check. I guess I did so to fix some broken test.

Regards`ValueSequence` it uses a common super-type. So if you have a `node()` and an `xs:string` in the value sequence, the common super type is `item()`... no?

**Comment by wolfgangmm** [comment]
See #1133, which introduced the changes in question to `PathExpr`. I'm fine to just drop code nobody can explain, but you asked about that particular line, so I thought it is not ok and we need a solution.

**Comment by wolfgangmm** [comment]
If the sequence is indeed mixed, it is correct to return `Type.ITEM` and in this case the check in `PathExpr` should fail as it would.

**Comment by adamretter** [comment]
@wolfgangmm I just spotted a problem with your suggestion to change `SequenceType#getItemType()` to `return sequence.getItemType();`.

When given code like:
```xquery
let $seq := (<a/>, <b/>, 1, 2)
return
  subsequece($seq, 1,2)
```
Your change would mean that we return `Type.ITEM` from the underlying value sequence, but the subsequence should actually return `Type.NODE` (or even `Type.ELEMENT`).

We cannot infer the type from the subequence without actually iterating all of the items in the subsequence and checking their common super type. We really want to avoid that, as a lazy approach is much more performant and memory kind.

Thoughts?

**Comment by adamretter** [comment]
> So even if you have to iterate, you do not need to continue beyond item 2, which means the performance impact should be small?

heh. Sure that works for a small subsequence, but what about a sequence of 20,000 items, where you want the first `999` ?

**Comment by wolfgangmm** [comment]
> heh. Sure that works for a small subsequence, but what about a sequence of `20,000` items I , where you want the first `999` ?

In this case you only need to determine the type of the first 999, so there's no performance loss, right?

> I guess I am trying to figure out if we _have_ to take determine the type as you are suggesting, or if you think that the fix I proposed previously has a problem?

The whole issue was that using `subsequence` as a path step *should* result in an error if atomic values are returned. At least this is why `PathExpr` tries to determine the type. So `subsequence((<a>a</a>, "a"), 1, 2)/text()` would be an error. We could argue that those checks should actually be done elsewhere and it's not the job of subsequence to determine the type. 

However, I think having correct type information returned by expressions where possible is always preferable to not having the information.

**Comment by wolfgangmm** [comment]
I guess what I'm thinking is: if there is an easy and fast way to determine the type of the result of subsequence, we should rather go for it instead of having to modify the superclass `Function` and `PathExpr`.

But if it means a performance loss, I'm rather for speed.

**Comment by adamretter** [comment]
@wolfgangmm I have commited the change you suggested. It looks good. Thank you :-)

**Pull request merged**